File: robot-bounded-in-circle/Solution1041.java
// https://leetcode.com/problems/robot-bounded-in-circle/

class Solution1041 {

    static enum Dir {
        North, South, East, West
    };

    // T: O(n), S: O(1)
    public boolean isRobotBounded(String instructions) {
        int x = 0, y = 0;
        var dir = Dir.North;
        for (char d : instructions.toCharArray()) {
            if (d == 'G') {
                if (dir == Dir.North) {
                    y++;
                } else if (dir == Dir.South) {
                    y--;
                } else if (dir == Dir.East) {
                    x++;
                } else if (dir == Dir.West) {
                    x--;
                }
            } else if (d == 'L') {
                if (dir == Dir.North) {
                    dir = Dir.West;
                } else if (dir == Dir.South) {
                    dir = Dir.East;
                } else if (dir == Dir.East) {
                    dir = Dir.North;
                } else if (dir == Dir.West) {
                    dir = Dir.South;
                }
            } else if (d == 'R') {
                if (dir == Dir.North) {
                    dir = Dir.East;
                } else if (dir == Dir.South) {
                    dir = Dir.West;
                } else if (dir == Dir.East) {
                    dir = Dir.South;
                } else if (dir == Dir.West) {
                    dir = Dir.North;
                }
            }
        }

        return (x == 0 && y == 0) || dir != Dir.North;
    }

    public static void main(String[] args) {
        var sol = new Solution1041();
        var instructions1 = "GGLLGG";
        var instructions2 = "GG";
        var instructions3 = "GL";
        System.out.println(sol.isRobotBounded(instructions1));
        System.out.println(sol.isRobotBounded(instructions2));
        System.out.println(sol.isRobotBounded(instructions3));
    }

}

File: robot-bounded-in-circle/solution1041.py
# https://leetcode.com/problems/robot-bounded-in-circle/

from enum import Enum


class Dir(Enum):
    North = 1
    South = 2
    East = 3
    West = 4


class Solution1041:

    # T: O(n), S: O(1)
    def isRobotBounded(self, instructions: str) -> bool:
        x, y = 0, 0
        dir = Dir.North
        for d in instructions:
            if d == 'G':
                if dir == Dir.North:
                    y += 1
                elif dir == Dir.South:
                    y -= 1
                elif dir == Dir.East:
                    x += 1
                elif dir == Dir.West:
                    x -= 1
            elif d == 'L':
                if dir == Dir.North:
                    dir = Dir.West
                elif dir == Dir.South:
                    dir = Dir.East
                elif dir == Dir.East:
                    dir = Dir.North
                elif dir == Dir.West:
                    dir = Dir.South
            elif d == 'R':
                if dir == Dir.North:
                    dir = Dir.East
                elif dir == Dir.South:
                    dir = Dir.West
                elif dir == Dir.East:
                    dir = Dir.South
                elif dir == Dir.West:
                    dir = Dir.North

        return (x == 0 and y == 0) or dir != Dir.North


def main() -> None:
    sol = Solution1041()
    instructions1 = "GGLLGG"
    instructions2 = "GG"
    instructions3 = "GL"
    print(sol.isRobotBounded(instructions1))  # True
    print(sol.isRobotBounded(instructions2))  # False
    print(sol.isRobotBounded(instructions3))  # True


if __name__ == "__main__":
    main()
